include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -78,195 +193,517 @@ template <typename> class FullMatrix;
+               MatrixArray &&derivative_matrices_)
 {
-  Assert(size == -1 ||
-         (size > 0 && static_cast<unsigned int>(size) == mass_matrix.m()),
-         ExcDimensionMismatch(size, mass_matrix.m()));
-  AssertDimension(mass_matrix.m(), mass_matrix.n());
-  AssertDimension(mass_matrix.m(), derivative_matrix.m());
-  AssertDimension(mass_matrix.m(), derivative_matrix.n());
-
-  this->mass_matrix = mass_matrix;
-  this->derivative_matrix = derivative_matrix;
-
-  std::vector<Vector<Number> > eigenvecs(mass_matrix.m());
-  LAPACKFullMatrix<Number> mass_copy(mass_matrix.m(), mass_matrix.n());
-  LAPACKFullMatrix<Number> deriv_copy(derivative_matrix.m(), derivative_matrix.n());
-  mass_copy = mass_matrix;
-  deriv_copy = derivative_matrix;
-
-  deriv_copy.compute_generalized_eigenvalues_symmetric(mass_copy, eigenvecs);
-  AssertDimension(eigenvecs.size(), mass_matrix.m());
-  eigenvectors.reinit(mass_matrix.m(), mass_matrix.m());
-  for (unsigned int i=0; i<mass_matrix.m(); ++i)
-    for (unsigned int j=0; j<mass_matrix.n(); ++j)
-      eigenvectors(i,j) = eigenvecs[j][i];
-
-  eigenvalues.resize(mass_matrix.m());
-  for (unsigned int i=0; i<mass_matrix.m(); ++i)
-    eigenvalues[i] = deriv_copy.eigenvalue(i).real();
+  auto &&mass_matrices = std::forward<MatrixArray>(mass_matrices_) ;
+  auto &&derivative_matrices = std::forward<MatrixArray>(derivative_matrices_) ;
+  this->mass_matrix = mass_matrices ;
+  this->derivative_matrix = derivative_matrices ;
+
+  for (int dir = 0; dir < dim; ++dir)
+    {
+      Assert (size == -1 || (size > 0 && static_cast<unsigned int>(size) == mass_matrices[dir].n_rows()),
+              ExcDimensionMismatch(size, mass_matrices[dir].n_rows()));
+      AssertDimension (mass_matrices[dir].n_rows(), mass_matrices[dir].n_cols());
+      AssertDimension (mass_matrices[dir].n_rows(), derivative_matrices[dir].n_rows());
+      AssertDimension (mass_matrices[dir].n_rows(), derivative_matrices[dir].n_cols());
+
+      this->eigenvectors[dir].reinit (mass_matrices[dir].n_cols(), mass_matrices[dir].n_rows()) ;
+      this->eigenvalues[dir].resize (mass_matrices[dir].n_cols()) ;
+      spectral_assembly<Number> (&(mass_matrices[dir](0,0))
+                                 , &(derivative_matrices[dir](0,0))
+                                 , mass_matrices[dir].n_rows()
+                                 , mass_matrices[dir].n_cols()
+                                 , this->eigenvalues[dir].begin()
+                                 , &(this->eigenvectors[dir](0,0))) ;
+    }
 }
 
 
 
 template <int dim, typename Number, int size>
 inline
-unsigned int
-TensorProductMatrixSymmetricSum<dim,Number,size>::m() const
+void
+TensorProductMatrixSymmetricSum<dim,Number,size>
+::reinit (const std::array<Table<2,Number>, dim> &mass_matrix,
+          const std::array<Table<2,Number>, dim> &derivative_matrix)
 {
-  return Utilities::fixed_power<dim>(mass_matrix.m());
+  reinit_impl (mass_matrix, derivative_matrix) ;
```

Comment:
One more style issue: You sometimes add a space between the final semicolon and sometimes not. In deal.II we typically do not. Could you please unify the style?